### PR TITLE
Make `vec::IntoIter` covariant again

### LINKED ISSUE #35721: vec::IntoIter is now invariant
#35447 replaced a `*const T` by a `*mut T` in `IntoIter<T>`, which changed `IntoIter` from being covariant to being invariant. This breaks crates like `vec_map` which relied on `IntoIter` being invariant, e.g. https://travis-ci.org/asajeffrey/presort/jobs/152720189#L208 has error:
```
error[E0308]: mismatched types
   --> /home/ajeffrey/.cargo/registry/src/github.com-1ecc6299db9ec823/vec_map-0.6.0/src/lib.rs:952:84
    |
952 |     fn into_iter_covariant<'a, T>(iter: IntoIter<&'static T>) -> IntoIter<&'a T> { iter }
    |                                                                                    ^^^^ lifetime mismatch
    |
    = note: expected type `IntoIter<&'a T>`
    = note:    found type `IntoIter<&'static T>`
note: the lifetime 'a as defined on the block at 952:81...
   --> /home/ajeffrey/.cargo/registry/src/github.com-1ecc6299db9ec823/vec_map-0.6.0/src/lib.rs:952:82
    |
952 |     fn into_iter_covariant<'a, T>(iter: IntoIter<&'static T>) -> IntoIter<&'a T> { iter }
    |                                                                                  ^^^^^^^^
    = note: ...does not necessarily outlive the static lifetime

   Compiling rustc-serialize v0.3.19
error: aborting due to previous error
```
cc @alexcrichton @frewsxcv @apasel422

### PR DESCRIPTION
**Pull request by apasel422**
Closes #35721 

r? @alexcrichton

### CODE REVIEW

**Comment by apasel422** [comment]
This uses integer indices instead of pointers to avoid aliasing `RawBuf`'s underlying `Unique`, but could be rewritten, albeit with some duplication from `RawBuf`, to use a struct layout like

```rust
struct IntoIter<T> {
  buf: Shared<T>,
  cap: usize,
  head: Shared<T>,
  tail: Shared<T>,
}
```

File: src/libcollections/vec.rs [hunk 1]
```diff
@@ -1812,27 +1786,22 @@ impl<T> DoubleEndedIterator for IntoIter<T> {
     #[inline]
     fn next_back(&mut self) -> Option<T> {
         unsafe {
-            if self.end == self.ptr {
+            if self.head == self.tail {
```
Comment (frewsxcv) [inline src/libcollections/vec.rs]: This could also be: `if self.is_empty()`, I think?

**Comment by alexcrichton** [comment]
In the past this has been pretty performance-sensitive, could you benchmark just a forward iteration of a `IntoIter` to ensure that it didn't regress? Other than that looks great to me, thanks!

Also nominating for backport as beta was cut today and we don't want to miss this.

**Comment by apasel422** [comment]
@alexcrichton Ah, if someone else could help out with benchmarks, that'd be great. If they indicate that this approach is slower, I'd be happy to rewrite this to use the other approach I mentioned.

**Comment by alexcrichton** [comment]
@apasel422 could you gist the IR for this:

```rust
extern {
    fn bar(a: u8);
}
pub fn sum(v: Vec<u8>) {
    for e in v.into_iter() {
        unsafe { bar(e); }
    }
}
```

For me, `rustc bar.rs -O --crate-type lib --emit llvm-ir` looks like [this](https://gist.github.com/alexcrichton/a3f19f943356d2fd9eef986c41f7cc41). I think we basically just need to be sure that this still optimizes to a tight loop.

**Comment by apasel422** [comment]
IR Gist: https://gist.github.com/apasel422/c16aac732979a21c5a69c5180ca434a2

**Comment by alexcrichton** [comment]
Hm so that's *different* because LLVM still [has an add instruction](https://gist.github.com/apasel422/c16aac732979a21c5a69c5180ca434a2#file-foo-ll-L46), I'm not sure how much that'll end up translating to missed optimizations though. Maybe it's worth to stick to a two-pointers-approaching-each-other implementation though?

**Comment by apasel422** [comment]
@alexcrichton Done.

**Comment by alexcrichton** [comment]
@bors: r+ 7e148cd06292d125f925323fda6fabbc625a42c8

Thanks! (also wow that was fast)

**Comment by retep998** [comment]
Might be worth backporting this to beta, as the regression lived just long enough to jump into beta.

**Comment by apasel422** [comment]
@alexcrichton Can we get this backported to beta?

**Comment by alexcrichton** [comment]
Yes this is tagged with beta-nominated, and we'll discuss at the next libs triage meeting. It'll likely be an obvious "yes let's backport"

**Comment by alexcrichton** [comment]
Discussion at @rust-lang/libs today decided to accept for backport

**Pull request merged**